### PR TITLE
ACCUMULO-4667 Reworked the LocalityGroupIterator to more efficiently …

### PR DESCRIPTION
**Pull request by ivakegg**
…determine which groups to search when seeking.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 1]
```diff
@@ -97,75 +133,118 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
...
+      if (groups.defaultGroup != null) {
         if (inclusive) {
-          if (!nonDefaultColumnFamilies.containsAll(cfSet)) {
+          if (!groups.groupByCf.keySet().containsAll(cfSet)) {
             // default LG may contain wanted and unwanted column families
-            include = true;
+            groupsToUse.add(groups.defaultGroup);
           }// else - everything wanted is in other locality groups, so nothing to do
         } else {
           // must include, if all excluded column families are in other locality groups
```
Comment (brianloss) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: I know this wasn't your change, but this comment is misleading. We haven't tested that all excluded column families are in other locality groups here. I think that for !inclusive, you have to use the default locality group no matter what.

**Comment by keith-turner** [comment]
I created a [little program](https://gist.github.com/keith-turner/18aeadfad1f46e919cd4649fa59b3787) to experiment with the changes in this branch.  I am seeing significantly improved performance (this PR vs master) when fetching two columns from a file with 10K column families across 10 loc groups.  I plan to do more experimentation and review the code tomorrow.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 2]
```diff
@@ -97,75 +133,117 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
...
 
-      if (cfSet.size() == 0) {
-        include = !inclusive;
-      } else if (lgr.isDefaultLocalityGroup && lgr.columnFamilies == null) {
-        // do not know what column families are in the default locality group,
-        // only know what column families are not in it
+    // if no column families specified, then include all groups unless !inclusive
+    if (cfSet.size() == 0) {
+      if (!inclusive) {
+        groupsToUse.addAll(Arrays.asList(groups.groups));
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: For this case we could have an immutable set that is always used.  Create this once in `LocalityGroupContext`.  Then we can avoid allocating the `HashSet` for `groupsToUse`, just allocate it in the else.  We can also avoid populating the hashset.  Could use Guava immutable set

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 3]
```diff
@@ -97,75 +133,117 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
...
+       * Need to consider the following cases for inclusive and exclusive (lgcf:locality group column family set, cf:column family set) lgcf and cf are disjoint
+       * lgcf and cf are the same cf contains lgcf lgcf contains cf lgccf and cf intersect but neither is a subset of the other
+       */
+      if (!inclusive) {
+        for (Entry<ByteSequence,LocalityGroup> entry : groups.groupByCf.entrySet()) {
+          if (!cfSet.contains(entry.getKey())) {
+            groupsToUse.add(entry.getValue());
+          }
+        }
+      } else if (groups.groupByCf.size() <= cfSet.size()) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: This is very nice!

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 4]
```diff
@@ -97,75 +133,117 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
...
+        lgr.getIterator().seek(range, EMPTY_CF_SET, false);
+        addSource(lgr.getIterator());
+      }
+    } else {
+      if (lastUsed == null) {
+        lastUsed = new ArrayList<LocalityGroup>(lgContext.groups.length);
+      } else {
+        lastUsed.clear();
+      }
+      lastColumnFamilies = cfSet;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: Need to be careful with this, it could be a ref to a set a user passed.  The user could call `seek(setA)` then modify `setA` and then call `seek(setA)`.  In this use case it seems like the wrong loc groups could be used.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 2 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: Good observation.  Thanks.  Will do.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 3 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: Thank you

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 4 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: good point....I will try and rework this part.

**Comment by milleruntime** [comment]
@ivakegg I recently created a performance test suite using JMH to measure the work I did with iterators: https://github.com/milleruntime/accumulo-benchmark
It is pretty specific but could be useful for measuring your improvements.

**Comment by keith-turner** [comment]
Squashing commits while the review is in progress makes it hard to know what changed.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 5]
```diff
@@ -97,75 +133,116 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
...
 
     return numLGSeeked;
   }
 
   @Override
   public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
-    seek(this, groups, nonDefaultColumnFamilies, range, columnFamilies, inclusive);
+    Set<ByteSequence> cfSet;
+    if (columnFamilies.size() > 0)
+      cfSet = new HashSet<>(columnFamilies);
```
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: I like this idea.  The change to make the cfset an ImmutableSet in the first place is trivial in the LocalityGroupUtil.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 6]
```diff
@@ -97,75 +134,122 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
...
   }
 
   @Override
   public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
-    seek(this, groups, nonDefaultColumnFamilies, range, columnFamilies, inclusive);
+
+    // determine if the arguments have changed since the last time
+    boolean sameArgs = false;
+    ImmutableSet<ByteSequence> cfSet = null;
+    if (lastUsed != null && inclusive == lastInclusive) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: I was curious if the RFile unit test would cover these different branches.  So I ran those test with code coverage.  None of the code in this function was touched.  I was surprised and dug into it, this function is only used by the iin memory map.  Rfile only calsee the static seek method in this code. 

It would be nice to test that seeking the same iterator with different families and inclusiveness works.

Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: @keith-turner OK, I will work on that....

Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: It appears that the InMemoryMapTest.testLocalityGroups test does go through this logic and indeed changes the column families and changes the inclusiveness testing all of these branches.  I am not sure how the code coverage missed this.  Can you verify please.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 7]
```diff
@@ -16,23 +16,28 @@
  */
 package org.apache.accumulo.core.iterators.system;
 
+import com.google.common.collect.ImmutableSet;
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: Can you please revert the reordering of imports?

If the imports are actually in the wrong order, please open up another issue and lets squash them all at once.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 8]
```diff
@@ -65,23 +70,55 @@ public InterruptibleIterator getIterator() {
...
 
   @Override
   public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
     throw new UnsupportedOperationException();
   }
 
-  public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<ByteSequence> nonDefaultColumnFamilies, Range range,
-      Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
+  public static final int seek(HeapIterator hiter, LocalityGroupContext groups, Range range, Collection<ByteSequence> columnFamilies, boolean inclusive,
+      List<LocalityGroup> used) throws IOException {
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: What is `used`? Could you add some javadoc on this method, please?

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 9]
```diff
@@ -97,75 +134,122 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
     else
       cfSet = Collections.emptySet();
 
-    for (LocalityGroup lgr : groups) {
-      // when include is set to true it means this locality groups contains
-      // wanted column families
-      boolean include = false;
+    // determine the set of groups to use
+    Collection<LocalityGroup> groupsToUse = Collections.EMPTY_LIST;
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: `Collections.emptyList()` instead? From the method javadoc:

> Using this method is likely to have comparable cost to using the like-named field.  (Unlike this method, the field does not provide type safety.)

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java [hunk 10]
```diff
@@ -97,75 +134,122 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
...
+
+    // determine if the arguments have changed since the last time
+    boolean sameArgs = false;
+    ImmutableSet<ByteSequence> cfSet = null;
+    if (lastUsed != null && inclusive == lastInclusive) {
+      if (columnFamilies instanceof Set) {
+        sameArgs = lastColumnFamilies.equals(columnFamilies);
+      } else {
+        cfSet = ImmutableSet.copyOf(columnFamilies);
+        sameArgs = lastColumnFamilies.equals(cfSet);
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: `sameArgs` is essentially computed the same way in both branches. Could pull it out.

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java [hunk 11]
```diff
@@ -50,16 +52,16 @@
 
   // private static final Logger log = Logger.getLogger(ColumnFamilySet.class);
 
-  public static final Set<ByteSequence> EMPTY_CF_SET = Collections.emptySet();
+  public static final Set<ByteSequence> EMPTY_CF_SET = ImmutableSet.of();
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: Why this change? Isn't the `Set` returned by `Collections.emptySet()` also immutable?

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java [hunk 12]
```diff
@@ -50,16 +52,16 @@
...
     if (columns.size() == 0)
       return EMPTY_CF_SET;
-    Set<ByteSequence> result = new HashSet<>(columns.size());
+    Builder<ByteSequence> builder = ImmutableSet.builder();
     for (Column col : columns) {
-      result.add(new ArrayByteSequence(col.getColumnFamily()));
+      builder.add(new ArrayByteSequence(col.getColumnFamily()));
     }
-    return result;
+    return builder.build();
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: Could make this a bit more concise with a Java8 lambda since this is slated for master.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 6 above)
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: @ivakegg  my memory is a bit fuzzy on this.  Its good that InMemoryMapTest covers these changes.  However, I think some of these optimizations you made will benefit iterators seeking the in mem map, but not iterators seeking RFiles.  I think I was not surprised at lack of coverage, but at the lack of use of your optimizations by RFile.  Personally I would like to see RFile refactored to use your optimizations.

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java (hunk 11 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: To take advantage of the ImmutableSet comparison logic, I needed this set to also be an ImmutableSet.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 10 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: To take advantage of the ImmutableSet comparison logic, I created an ImmutableSet here....

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java (hunk 12 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: I will look into that, although I am hoping that this can be backported to 1.8.2 as well.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 6 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: Ah, I see what you were getting at.  I will take a closer look at the RFile path....

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 7 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: I can reset the order.  Why another Issue?

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 8 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: ok

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 9 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: ok

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java (hunk 11 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: What ImmutableSet comparison logic? re-skimming your patch, it's not jumping out at me.

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java (hunk 12 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: Cool, no worries.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 7 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: We have a well-defined ordering of imports in Accumulo. I suggested that if these were re-ordered to adhered to that ordering, let's do that, just in a separate issue.

If the ordering was your just IDE moving stuff around, revert here is all we need.

Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: I believe it was only my IDE in this case....

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java (hunk 11 above)
Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: You have to dig into the ImmutableSet.equals method to see the efficiencies it has.  Keith put me onto this.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 7 above)
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: I actually don't think we have a well-defined ordering in Accumulo. If we do, I don't know what it is. I think we've mostly used Eclipse defaults from the beginning, but as other developers use other IDEs, it's probably important that we establish a well-defined ordering (if we don't have one).

I know Fluo has a well-defined ordering, but it's hard to enforce automatically. I'm [currently writing a Maven plugin to automatically sort imports in the build][1]. So that way, this won't be an issue at all. I've been testing it with Accumulo, and it works well, but it's slow. It adds 2.5 minutes to the build time on Accumulo. I'm working on migrating some of the parsing code I wrote using RegExs to a proper AST parser, which should be about 10x faster, but there's some fringe cases that make that hard. In particular, Java files can have "orphaned" comments (not connected to any other node in the parse tree) interspersed with imports, which are annoying to deal with. And also, the libraries I've looked at don't seem to make it easy to rewrite a file's imports without changing the rest of the file's style.

[1]: https://github.com/revelc/impsort-maven-plugin

File: core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java (hunk 11 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: Oh, like the reference check against `this`? I could see that being beneficial.

A comment in the code as to use for ImmutableSet for that reason would be quite appreciated.

Comment (ivakegg) [inline core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java]: Right you are.  I shall add comments.  Thanks.

File: core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java (hunk 7 above)
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java]: Oh, oops. I thought we had one, but I could see it just being eclipse's default. At least one other project I'm a part of uses https://google.github.io/styleguide/javaguide.html#s3.3-import-statements which I've grown to like.

Anyways, this is becoming quite the tangent. Thanks for changing it back, Ivan.

**Comment by keith-turner** [comment]
@ivakegg  I like how you refactored LocalityGroupIterator so RFile could call the efficient seek code.

**Comment by ivakegg** [comment]
@keith-turner Thank you kind sir

**Comment by ivakegg** [comment]
@keith-turner I did forget one thing: I need to ensure that ImmutableSet is used for the columnFamilies through the RFile path....I will update as such

**Comment by ivakegg** [comment]
@keith-turner Nevermind....since the cfSet from the top level is an ImmutableSet, then this should work fine....

**Comment by ivakegg** [comment]
@keith-turner Awesome.  I am now leaning towards merging this one in if no one has any objections.
